src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CSupport.java

```diff
@@ -49,6 +50,163 @@ public static ForeignLinker getSystemLinker() {
         return SharedUtils.getSystemLinker();
     }
 
+    public static VaList newVaList(Consumer<VaList.Builder> actions) {
+        return SharedUtils.newVaList(actions);
+    }
+
+    /**
+     * An interface that models a C {@code va_list}.
+     *
+     * Per the C specification (see C standard 6.5.2.2 Function calls - item 6),
+     * arguments to variadic calls are erased by way of 'default argument promotions',
+     * which erases integral types by way of integer promotion (see C standard 6.3.1.1 - item 2),
+     * and which erases all {@code float} arguments to {@code double}.
+     *
+     * As such, this interface only supports reading {@code int}, {@code double},
+     * and any other type that fits into a {@code long}.
+     */
+    public interface VaList extends AutoCloseable {
+
+        /**
+         * Reads a value into an {@code int}
+         *
+         * @param layout the layout of the value
+         * @return the value read as an {@code int}
+         */
+        int readInt(MemoryLayout layout);
```

Comment:
Here some confusion might arise as to what `int` means. `int` means Java `int`.
As discussed offline, it would be better to use some hint e.g. `from` to indicate that we are converting *from* a Java value to a native one.
Also, `read` for some reason is reminiscent of I/O.